vault/token_store.go

```diff
@@ -1010,6 +1022,85 @@ func (ts *TokenStore) lookupBySaltedAccessor(saltedAccessor string) (accessorEnt
 	return aEntry, nil
 }
 
+// handleCleanup handles the cleaning up of leaked accessor storage entries and
+// cleaning up of leases that are associated to tokens that are expired.
+func (ts *TokenStore) handleCleanup(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	// List out all the accessors
+	saltedAccessorList, err := ts.view.List(accessorPrefix)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch accessor entries: %v", err)
+	}
+
+	var cleanupErrors *multierror.Error
+
+	// For each of the accessor, see if the token ID associated with it is
+	// a valid one. If not, delete the leases associated with that token
+	// and delete the accessor as well.
+	for _, saltedAccessor := range saltedAccessorList {
+		accessorEntry, err := ts.lookupBySaltedAccessor(saltedAccessor)
+		if err != nil {
+			cleanupErrors = multierror.Append(cleanupErrors, fmt.Errorf("failed to read the accessor entry: %v", err))
+			continue
+		}
+
+		// A valid accessor storage entry should always have a token ID
+		// in it. If not, it is an invalid accessor entry and needs to
+		// be deleted.
+		if accessorEntry.TokenID == "" {
+			// If deletion of accessor fails, move on to the next
+			// item since this is just a best-effort operation
+			err = ts.view.Delete(accessorPrefix + saltedAccessor)
+			if err != nil {
+				cleanupErrors = multierror.Append(cleanupErrors, fmt.Errorf("failed to delete the accessor entry: %v", err))
+				continue
+			}
+		}
+
+		te, err := ts.lookupSalted(ts.SaltID(accessorEntry.TokenID))
+
+		// If token entry is not fetched for whatever reason, assume
+		// that the token is no more valid and conclude that accessor
+		// for this token should not exist, the leases associated with
+		// the token should not exist as well. Also, in case token
+		// entry exists, and if NumUses in the token entry is marked as
+		// -3, it a hint that the previous revocation operation had
+		// failed. In that case, attempt a retry operation.
+		if err != nil || te == nil || (te != nil && te.NumUses == -3) {
```

Comment:
Are you sure that the right thing to do in an *error* case is perform this revocation? An error case is likely to be some kind of underlying storage error, I think; not finding the entry should simply return nil.